apis/v1alpha2/validation/httproute.go

```diff
@@ -166,3 +166,29 @@ func validateHTTPPathMatch(path *gatewayv1a2.HTTPPathMatch, fldPath *field.Path)
 	}
 	return allErrs
 }
+
+// validateHTTPRouteFilterTypeMatchesValues validates that the filter type matches the filter value.
+func validateHTTPRouteFilterTypeMatchesValues(filter *gatewayv1a2.HTTPRouteFilter, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+
+	switch filter.Type {
+	case gatewayv1a2.HTTPRouteFilterExtensionRef:
+		if filter.ExtensionRef == nil || filter.RequestHeaderModifier != nil || filter.RequestMirror != nil || filter.RequestRedirect != nil {
```

Comment:
These if logics might not so feasible, if a new filter type is introduced, we have to change a few places, and it's easy to forget.

Reply:
I see what you mean. It's hard to mantain. I've researched alternatives:
1. I started by looking at ways to implement discriminated unions in Go and they usually point to using interfaces(see [Alternatives to sum types in Go](https://making.pusher.com/alternatives-to-sum-types-in-go/), [StackOverflow](https://stackoverflow.com/questions/21553398/best-practice-for-unions-in-go)). Correct me If I'm wrong but I don't think it's right to add interfaces to the API just to solve this problem. 
2. Something that occurred to me is to compare the json tags with the type but I think is  "hacky".

Do you have any ideas or tips for the correct implementation? I have a hunch that there is a more elegant way to solve this and I just didn't look in the right place.

Reply:
If we can relax the constraints, say unused filter config can still be reserved there, than the switch statement can be simple.

Reply:
The best I could come up with was this:

```Go
func validateUnionField(discriminator string, name string, value interface{}) {
        var isNil bool

        switch value := value.(type) {
        case *gatewayv1a2.HTTPRequestHeaderFilter:   
                isNil = (value == nil)
        case *gatewayv1a2.HTTPRequestMirrorFilter:
                isNil = (value == nil)
        case *gatewayv1a2.HTTPRequestRedirectFilter:
                isNil = (value == nil)
        case *gatewayv1a2.LocalObjectReference:
                isNil = (value == nil)
        default:
                // Puke an error so someone updates this code ...
        
        }

        if discriminator == name {
                if isNil {               
                        // Filter spec is not set, but it should be.
                }
        } else {
                if !isNil {
                        // Filter spec is set, but it should not be.
                }
        }
}

validateUnionField(filter.Type, gatewayv1a2.HTTPRouteFilterExtensionRef, filter.ExtensionRef)
validateUnionField(filter.Type, gatewayv1a2.HTTPRouteFilterRequestHeaderModifier, filter.RequestHeaderModifier)
         
```

Reply:
Made the changes. Does this look good to you?

Reply:
I've relaxed the constraints to make the switch statement simpler. I think that `validateUnionField`  is more complex and still has the issue of referencing the types so new types need to be added.